python/pyspark/sql/types.py

```diff
@@ -997,12 +997,13 @@ def _restore_object(dataType, obj):
     # same object in most cases.
     k = id(dataType)
     cls = _cached_cls.get(k)
-    if cls is None:
+    if cls is None or cls.__datatype is not dataType:
```

Comment:
I think we should change it into an assert, since that will make it easier to debug this issue if it ever crops up again (if we accidentally removed the reference keeping the `DataType` alive, for example).